### ISSUE TITLE
Race: shoot system components are missing `KUBERNETES_SERVICE_HOST` env var on cluster creation

### ISSUE THREAD

**Issue by timebertt** [gardener/gardener#7259]
**How to categorize this issue?**

/area robustness
/kind bug

**What happened**:

On shoot creation, some `kube-system` pods were created before the `apiserver-proxy.networking.gardener.cloud` webhook.
With this, the pods did not get the `KUBERNETES_SERVICE_HOST` env var injected ~~and can't connect to the shoot API server (because they are talking directly to the `kubernetes.default.svc` instead of using the `apiserver-proxy`).~~

This led to pods being unable to connect to the shoot API server in combination with a bug in the ACL extension (ref https://github.com/stackitcloud/gardener-extension-acl/pull/19, [full explanation](https://github.com/gardener/gardener/issues/7259#issuecomment-1369475974)).

```
$ ks get po
NAME                                                 READY   STATUS             RESTARTS          AGE
apiserver-proxy-cgdtt                                2/2     Running            0                 24h
calico-kube-controllers-55c8c77-8kb5j                0/1     CrashLoopBackOff   280 (3m48s ago)   24h
calico-node-ghm9l                                    1/1     Running            0                 24h
calico-node-vertical-autoscaler-754b94fb77-mdc7w     0/1     CrashLoopBackOff   290 (31s ago)     24h
calico-typha-deploy-6dd875bb76-jjh7j                 1/1     Running            0                 24h
calico-typha-horizontal-autoscaler-8848ccc84-5czvn   1/1     Running            0                 24h
calico-typha-vertical-autoscaler-86bc8c4685-54dx7    0/1     CrashLoopBackOff   289 (2m5s ago)    24h
coredns-66f96f675f-44dtp                             1/1     Running            0                 24h
coredns-66f96f675f-p7ckf                             1/1     Running            0                 24h
csi-driver-node-9k9fc                                3/3     Running            0                 24h
kube-proxy-pool-72mx7nl1is-v1.23.12-82v5c            2/2     Running            0                 24h
metrics-server-85d5c7445-8f8zm                       1/1     Running            0                 24h
node-exporter-lmmwt                                  1/1     Running            0                 24h
node-problem-detector-x8z8x                          1/1     Running            0                 24h
vpn-shoot-84c4849889-rzf7r                           1/1     Running            1 (23h ago)       24h

$ k get mutatingwebhookconfiguration apiserver-proxy.networking.gardener.cloud -oyaml
apiVersion: admissionregistration.k8s.io/v1
kind: MutatingWebhookConfiguration
metadata:
# ...
  creationTimestamp: "2022-12-21T12:32:17Z"

$ ks get po calico-kube-controllers-55c8c77-8kb5j -oyaml
apiVersion: v1
kind: Pod
metadata:
# ...
  creationTimestamp: "2022-12-21T12:32:16Z"
# ...
status:
# ...
    state:
      waiting:
        message: back-off 5m0s restarting failed container=calico-kube-controllers
          pod=calico-kube-controllers-55c8c77-8kb5j_kube-system(4456c528-9c83-4f31-979d-54e973304332)
        reason: CrashLoopBackOff

$ ks get po calico-kube-controllers-55c8c77-8kb5j -oyaml | grep KUBERNETES_SERVICE_HOST

$ ks get po calico-node-ghm9l -oyaml | grep KUBERNETES_SERVICE_HOST
    - name: KUBERNETES_SERVICE_HOST
    - name: KUBERNETES_SERVICE_HOST
```

This situation cannot self-heal. However, it can easily be resolved by deleting the crashlooping pods (the new pods will pass the webhook and receive the `KUBERNETES_SERVICE_HOST` env var).

**What you expected to happen**:

All pods (workload and especially system pods) get the `KUBERNETES_SERVICE_HOST` env var injected even on cluster creation.

**How to reproduce it (as minimally and precisely as possible)**:

This is a race, as the `Deploying shoot network plugin` task (and similar) don't wait for the `apiserver-proxy.networking.gardener.cloud` webhook to be present.

~~We seem to be lucky in reproducing this on clusters that enable https://github.com/stackitcloud/gardener-extension-acl. However, we couldn't find any reason why this extension might increase the likelihood of running into this race.~~
This race has been present for a longer time but only revealed itself with the bug fixed in https://github.com/stackitcloud/gardener-extension-acl/pull/19.

**Anything else we need to know?**:

We compared the task timestamps of shoot creations that succeed with shoot creations that run into the described bug.
Even in the successful case, we can observe that the `Network` extension is deployed in the exact same second as the `shoot-core` `ManagedResource` (which contains the webhook).
It seems like this race is already present since a long time, but only reveals itself under certain circumstances these days.

**Environment**:

- Gardener version: `v1.53.4`
- Cloud provider or hardware configuration: STACKIT (provider-openstack)

**Comment by timebertt** [comment]
We discussed this upfront in https://gardener-cloud.slack.com/archives/C045DSWJZB9/p1671698302273789, where it was mentioned that @oliver-goetz wants to refactor the `apiserver-proxy` to a golang component as part of https://github.com/gardener/gardener/issues/2754.

We agreed, that the refactoring should be performed before fixing this bug, as it requires extracting the `apiserver-proxy` form the `shoot-core` helm chart.
Once this refactoring is done, we only need to introduce a new reconciliation task that waits for the `apiserver-proxy` (including its webhook) and corresponding task dependencies.

**Comment by timebertt** [comment]
To speed up the fix for this bug, we will take over the `apiserver-proxy` refactoring task of https://github.com/gardener/gardener/issues/2754. @oliver-goetz

**Comment by timebertt** [comment]
Update:
As correctly [pointed out](https://gardener-cloud.slack.com/archives/C045DSWJZB9/p1671723650656589?thread_ts=1671698302.273789&cid=C045DSWJZB9) by @timuthy, injecting the FQDN of the API server in the `KUBERNETES_SERVICE_HOST` env var is only an optimization but not required for talking to the API server.

If the env var is not injected (or not respected by the client), traffic goes to the `kubernetes.default.svc` which is handled by `apiserver-proxy`. This in turn uses the proxy protocol to talk to the istio ingressgateway which allows it to identify which API server the client wants to talk to (by the `clusterIP` of the `kube-apiserver` service in the seed).

The race described in this issue should not be problematic because pods should still be able to communicate with the API server even if the env var is not injected by using the fallback path via the `apiserver-proxy`.
However, the race is problematic in combination with https://github.com/stackitcloud/gardener-extension-acl, which wasn't handling the fallback path correctly yet. This is fixed by https://github.com/stackitcloud/gardener-extension-acl/pull/19.

**TL;DR:**
The root cause of our problem was not the race described here but a missing allow list of the pod CIDR in gardener-extension-acl.
The race was present before and should still be fixed to make full use of the optimization.
The race only revealed itself in combination with the ACL bug. Without the ACL extension, the pods were still able to use the fallback path via the `apiserver-proxy`.

**Comment by timebertt** [comment]
This is not resolved yet.

/reopen
/remove-lifecycle rotten

**Comment by acumino** [comment]
GRM now deploys `apiserver-proxy-pod-webhook` [ref](https://github.com/gardener/gardener/pull/7980), this step is done at the very start, all pods in the shoot are created only after the GRM is healthy. The race condition described in the issue above will not occur now.

/close
